File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution:
    def binarySearch(self, root, node, traversal):
        # search for root, and return the traversal
        if not root:
            return traversal
        else:
            traversal.append(root)
            if root.val == node.val:
                return traversal
            
            # add this to the path taken down
            if node.val < root.val:
                return self.binarySearch(root.left, node, traversal)
            elif node.val > root.val:
                return self.binarySearch(root.right, node, traversal)

    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        p_path = self.binarySearch(root, p, [])
        q_path = self.binarySearch(root, q, [])
        
        best = root
        print(p_path, q_path)
        for p_node, q_node in zip(p_path, q_path):
            print(p_node.val, q_node.val)
            
            if p_node.val == q_node.val:
                best = p_node

        return best

File: Data Structures & Algorithms/binary-search/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int middleIndex = (left + right) / 2;
            int middle = nums[middleIndex];


            if (target == middle) {
                return middleIndex;
            }
            else if (target < middle) {
                right = middleIndex - 1;
            }
            else {
                left = middleIndex + 1;
            }
        }
        
        return -1;
    }
};
